File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/execution/MatchesEvaluator.java
package org.opencds.cqf.cql.engine.elm.execution;

import org.opencds.cqf.cql.engine.execution.Context;

public class MatchesEvaluator extends org.cqframework.cql.elm.execution.Matches {

    public static Object matches(String argument, String pattern) {
        if (argument == null || pattern == null) {
            return null;
        }

        return argument.matches(pattern);
    }

    @Override
    protected Object internalEvaluate(Context context) {
        String argument = (String) getOperand().get(0).evaluate(context);
        String pattern = (String) getOperand().get(1).evaluate(context);

        return matches(argument, pattern);
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/MatchesEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.runtime.Boolean
import org.opencds.cqf.cql.engine.runtime.String
import org.opencds.cqf.cql.engine.runtime.Value
import org.opencds.cqf.cql.engine.runtime.toCqlBoolean

object MatchesEvaluator {
    @JvmStatic
    fun matches(argument: Value?, pattern: Value?): Boolean? {
        if (argument == null || pattern == null) {
            return null
        }

        if (argument is String && pattern is String) {

            return argument.matches(pattern.value.toRegex()).toCqlBoolean()
        }

        throw InvalidOperatorArgument(
            "Matches(String, String)",
            "Matches(${argument.typeAsString}, ${pattern.typeAsString})",
        )
    }
}
